src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -70,85 +72,88 @@ public void Execute(GeneratorExecutionContext context)
+            RazorGenerateForSourceTexts(razorContext.RazorFiles, context, projectEngine);
         }
 
-        private void GenerateViews(GeneratorExecutionContext context, RazorSourceGenerationContext razorContext, RazorProjectEngine projectEngine)
+        private void RazorGenerateForSourceTexts(IReadOnlyList<RazorInputItem> files, GeneratorExecutionContext context, RazorProjectEngine projectEngine)
         {
-            var files = razorContext.CshtmlFiles;
-
             if (files.Count == 0)
             {
                 return;
             }
 
-            var arraypool = ArrayPool<(string, SourceText)>.Shared;
+            var arraypool = ArrayPool<(string, SourceText?)>.Shared;
             var outputs = arraypool.Rent(files.Count);
 
-            context.AddSource($"{context.Compilation.AssemblyName}.UnifiedAssembly.Info", ProvideApplicationPartFactoryAttributeSourceText);
-
             Parallel.For(0, files.Count, GetParallelOptions(context), i =>
             {
-                var file = files[i];
-
-                var codeDocument = projectEngine.Process(projectEngine.FileSystem.GetItem(file.NormalizedPath, FileKinds.Legacy));
-                var csharpDocument = codeDocument.GetCSharpDocument();
-                for (var j = 0; j < csharpDocument.Diagnostics.Count; j++)
-                {
-                    var razorDiagnostic = csharpDocument.Diagnostics[j];
-                    var csharpDiagnostic = razorDiagnostic.AsDiagnostic();
-                    context.ReportDiagnostic(csharpDiagnostic);
-                }
-
-                var generatedCode = csharpDocument.GeneratedCode;
-                var hint = GetIdentifierFromPath(file.GeneratedOutputPath ?? file.NormalizedPath);
-                outputs[i] = (hint, SourceText.From(generatedCode, Encoding.UTF8));
+                outputs[i] = ResolveGeneratedSourceTextFromFile(files[i], projectEngine, context);
             });
 
             for (var i = 0; i < files.Count; i++)
             {
                 var (hint, sourceText) = outputs[i];
-                context.AddSource(hint, sourceText);
+                if (sourceText != null)
+                {
+                    context.AddSource(hint, sourceText);
+                }
             }
 
             arraypool.Return(outputs);
         }
 
-        private static void CodeGenerateRazorComponents(GeneratorExecutionContext context, RazorSourceGenerationContext razorContext, RazorProjectEngine projectEngine)
+        private static (string, SourceText?) ResolveGeneratedSourceTextFromFile(RazorInputItem file, RazorProjectEngine projectEngine, GeneratorExecutionContext context)
         {
-            var files = razorContext.RazorFiles;
+            var hint = GetIdentifierFromPath(file.NormalizedPath);
+            if (file.FileKind == FileKinds.Legacy)
+            {
+                hint = GetIdentifierFromPath(file.GeneratedOutputPath ?? file.NormalizedPath);
```

Comment:
Might be worthwhile to drop `GeneratedOutputPath`. It's not super useful any more, is it?

Reply:
Yeah, we can. Ends up working better because for hints produced from GeneratedOutputPath the filename for the generated file (when you do `EmitCompilerGeneratedFiles`) ends up being really long:

```
_Users_captainsafia_githubcom_dotnet_sdk_blah_blah_Components_Counter.g.cs
```

vs

```
_Components_Counter.razor
```